docs/user-defined-type-functions.md

```diff
@@ -0,0 +1,238 @@
+# User-Defined Type Functions
+
+## Summary
+
+The new Luau type inference engine (or type solver, for short) adds support for a system of built-in type functions that compute result types from provided argument types. This system is already being used to power support for overloadable operators, and the broad design philosophy is to include built-in type functions that correspond to built-in runtime-level operations in the language. To further enhance the expressivity of the type system, this RFC extends that system with a proposed design for user-defined type functions, allowing developers to define their own type functions in ordinary Luau code.
+
+## Motivation
+
+The primary motivation for introducing user-defined type functions is to increase the expressiveness of the type system and empower developers to create more type-safe abstractions in their libraries. Current limitations of the type system prevent programmers from constructing types in relational ways, e.g. producing a new table where the types of every property have changed in some predictable way (like being made optional). These limitations can prevent developers from writing precisely-typed abstracts in their libraries, leading to less type-safe abstractions that rely on the use of `any` casting or less-than-desirable structural changes to their designed APIs. By adding support for user-defined type functions, we aim to:
+- Enable more precise type definitions that can capture complex relations on types
+- Facilitate the creation of type-safe, reusable and composable libraries
+- Enhance Luau's support for type-level programming
+
+The expected outcome is a more versatile type system that can adapt to a wider range of programming patterns, improving the Luau developer experience.
+
+## Design
+
+Type functions will be defined with the following syntax:
+```luau
+type function f(...)
+    -- implementation of the type function
+end
+```
```

Comment:
Section on the syntax could use more details to answer questions about visibility, scoping and shadowing.

I assume it will be similar to types?
As in, all type functions defined in the current scope including parent scopes are visible regardless is they are defined above or below each other.
With types, shadowing is not allowed (so if we introduce a new built-in type function, it will cause type errors in code that happens to use the same name). Maybe this can be relaxed to only disallow shadowing in the single scope?
Local type definitions are allowed and so are the type functions? That will imply the VM type function set has to be more dynamic instead of just collecting all type functions from outer scope once.

For parsing purposes, the type functions are parsed with an empty variable environment, similar to how they are compiled to global accesses.

Reply:
Yeah, I can add this information to the RFC.

I think what I originally had in mind is: Every type alias and type function would be defined in an order-independent way (i.e. we parse them all just as we do ordinary functions, and then we build up a mapping from symbols to their definitions to be used by the type system in invoking them). This would suggest a scoping rule that is essentially "one global scope for types, plus any nested runtime scopes within type function bodies." I think this should match how the existing "type functions" work, i.e. when you write `type

Reply:
Oh, to be more precise about current type shadowing rules, you are not allowed to shadow built-in types, but you can shadow a type from the global/env scope (at Roblox, `type Instance = number` will be ok, this way we freely add new environment types).
Since there are only built-in type functions and not additional ones injected from definition files (yet?), shadowing of built-ins will always be an error.

Reply:
Really, I'd just like it to be consistent with everything type aliases already do, so I suppose from that perspective we should probably just say we'll support scoping and so forth in the same way. Consistent scoping rules are good, and nobody has time to keep track of weird idiosyncratic additional ones.